### PR TITLE
Handle yanked gems

### LINKED ISSUE #4344: Handle yanked gems in a clear way
Today, when a gem is yanked, we throw the exact same error as when you manually enter a version number into your Gemfile that doesn't actually exist. If a gem is yanked, we have evidence that the version used to exist (in the lockfile), and so we should be able to print an error message that is much more informative, something along the lines of: 

> Your bundle is locked to foo-1.2.3, but that version could not be found in any of the sources listed in your Gemfile. If you haven't changed sources, that means the author of foo-1.2.3 has removed it. You'll need to update your bundle to a different version of foo that hasn't been removed in order to install.

### PR DESCRIPTION
**Pull request by allenzhao**
Close #4344

### CODE REVIEW

File: spec/install/yanked_spec.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+require "spec_helper"
+
+describe "handle yanked" do
+  before(:each) do
+    build_repo4 do
+      build_gem "foo", "9.0.0"
+    end
+  end
+
+  it "throws and error when there is gem yanked" do
```
Comment (segiddins) [inline spec/install/yanked_spec.rb]: `an error`

File: spec/install/yanked_spec.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+require "spec_helper"
+
+describe "handle yanked" do
```
Comment (segiddins) [inline spec/install/yanked_spec.rb]: `context "when installing a bundle that includes yanked gems"`

File: lib/bundler/definition.rb [hunk 3]
```diff
@@ -137,8 +137,17 @@ def resolve_remotely!
...
+        begin
+          specs = resolve.materialize(Bundler.settings[:cache_all_platforms] ? dependencies : requested_dependencies)
+        rescue GemNotFound => e # Handle yanked gem
+          gem_name, gem_version = extract_gem_info(e)
+          locked_gem = @locked_specs[gem_name].last
+          raise if locked_gem.nil? || locked_gem.version.to_s != gem_version
+          raise GemNotFound, "Your bundle is locked to #{locked_gem}, but that version could not " \
+                             "be found in any of the sources listed in your Gemfile. If you haven't changed sources, " \
+                             "that means the author of #{locked_gem} has removed it. You'll need to update your bundle " \
+                             "to a different version of foo that hasn't been removed in order to install."
```
Comment (RochesterinNYC) [inline lib/bundler/definition.rb]: it says "different version of foo" in the error message. I think this should be the gem name, right?

File: spec/install/yanked_spec.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+require "spec_helper"
+
+context "when installing a bundle that includes yanked gems" do
+  before(:each) do
+    build_repo4 do
+      build_gem "foo", "9.0.0"
+    end
+  end
+
+  it "throws an error when there is gem yanked" do
```
Comment (RochesterinNYC) [inline spec/install/yanked_spec.rb]: "when the original gem version is yanked"

File: spec/install/yanked_spec.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+    install_gemfile <<-G
+        source "file://#{gem_repo4}"
+        gem "foo", "10.0.0"
+    G
+
+    expect(out).to include("Your bundle is locked to foo (10.0.0)")
+  end
+
+  it "throws the original error when manually enter a version number that doesn't exist" do
```
Comment (RochesterinNYC) [inline spec/install/yanked_spec.rb]: "when only gemfile specifies a gem version that doesn't exist"

**Comment by RochesterinNYC** [comment]
@allenzhao made a few comments. Other than those small changes, this LGTM.

File: lib/bundler/definition.rb (hunk 3 above)
Comment (allenzhao) [inline lib/bundler/definition.rb]: Oh, I forgot to change that

**Comment by RochesterinNYC** [comment]
@segiddins, can you r+ this or delegate me? Thanks.

**Comment by segiddins** [comment]
@homu delegate=RochesterinNYC

**Comment by RochesterinNYC** [comment]
@homu r+

**Pull request merged**